File: src/main/java/dev/slimevr/autobone/StatsCalculator.java
package dev.slimevr.autobone;

import com.jme3.math.FastMath;


/// This is a stat calculator based on Welford's online algorithm
/// https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm
public class StatsCalculator {

	private int count = 0;
	private float mean = 0f;
	private float M2 = 0f;

	public void reset() {
		count = 0;
		mean = 0f;
		M2 = 0f;
	}

	public void addValue(float newValue) {
		count += 1;
		float delta = newValue - mean;
		mean += delta / count;
		float delta2 = newValue - mean;
		M2 += delta * delta2;
	}

	public float getMean() {
		return mean;
	}

	public float getVariance() {
		if (count < 1) {
			return Float.NaN;
		}

		return M2 / count;
	}

	public float getSampleVariance() {
		if (count < 2) {
			return Float.NaN;
		}

		return M2 / (count - 1);
	}

	public float getStandardDeviation() {
		return FastMath.sqrt(getVariance());
	}
}

File: server/core/src/main/java/dev/slimevr/autobone/StatsCalculator.kt
package dev.slimevr.autobone

import kotlin.math.*

/**
 * This is a stat calculator based on Welford's online algorithm
 * https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford%27s_online_algorithm
 */
class StatsCalculator {
	private var count = 0
	var mean = 0f
		private set
	private var m2 = 0f

	fun reset() {
		count = 0
		mean = 0f
		m2 = 0f
	}

	fun addValue(newValue: Float) {
		count += 1
		val delta = newValue - mean
		mean += delta / count
		val delta2 = newValue - mean
		m2 += delta * delta2
	}

	val variance: Float
		get() = if (count < 1) {
			Float.NaN
		} else {
			m2 / count
		}
	val sampleVariance: Float
		get() = if (count < 2) {
			Float.NaN
		} else {
			m2 / (count - 1)
		}
	val standardDeviation: Float
		get() = sqrt(variance)
}
